mmdet/models/dense_heads/base_dense_head.py

```diff
@@ -14,10 +17,181 @@ def loss(self, **kwargs):
+                Default: True.
+
+        Returns:
+            Tensor: Labeled boxes in shape (n, 5), where the first 4 columns
+                are bounding box positions (tl_x, tl_y, br_x, br_y) and the
+                5-th column is a score between 0 and 1.
+        """
+        if score_factor_list[0] is None:
+            # e.g. Retina, FreeAnchor, etc.
+            with_score_factors = False
+        else:
+            # e.g. FCOS, PAA, ATSS, etc.
+            with_score_factors = True
+
+        cfg = self.test_cfg if cfg is None else cfg
+        img_shape = img_meta['img_shape']
+        nms_pre = cfg.get('nms_pre', -1)
+
+        mlvl_bboxes = []
+        mlvl_scores = []
+        mlvl_score_factor = []
+        for level_idx, (cls_score, bbox_pred, score_factor) in enumerate(
+                zip(cls_score_list, bbox_pred_list, score_factor_list)):
+            assert cls_score.size()[-2:] == bbox_pred.size()[-2:]
+            featmap_size_hw = cls_score.shape[-2:]
+            cls_score = cls_score.permute(1, 2,
+                                          0).reshape(-1, self.cls_out_channels)
+            if self.use_sigmoid_cls:
+                scores = cls_score.sigmoid()
+            else:
+                scores = cls_score.softmax(-1)
+            bbox_pred = bbox_pred.permute(1, 2, 0).reshape(-1, 4)
+            if with_score_factors:
+                score_factor = score_factor.permute(1, 2,
+                                                    0).reshape(-1).sigmoid()
+            else:
+                score_factor = torch.ones_like(bbox_pred[:, 0])
+
+            if 0 < nms_pre < scores.shape[0]:
+                # Get maximum scores for foreground classes.
+                if self.use_sigmoid_cls:
+                    max_scores, _ = (scores * score_factor[:, None]).max(dim=1)
+                else:
+                    # remind that we set FG labels to [0, num_class-1]
+                    # since mmdet v2.0
+                    # BG cat_id: num_class
+                    max_scores, _ = (scores[:, :-1] *
+                                     score_factor[:, None]).max(dim=1)
+                _, topk_inds = max_scores.topk(nms_pre)
+
+                priors = self.prior_generator.sparse_priors(
+                    topk_inds, featmap_size_hw, level_idx, scores.dtype,
+                    scores.device)
+
+                bbox_pred = bbox_pred[topk_inds, :]
+                scores = scores[topk_inds, :]
+                score_factor = score_factor[topk_inds]
+            else:
+                priors = self.prior_generator.single_level_grid_priors(
+                    featmap_size_hw, level_idx, scores.device)
+
+            bboxes = self.bbox_coder.decode(
+                priors, bbox_pred, max_shape=img_shape)
+            mlvl_bboxes.append(bboxes)
+            mlvl_scores.append(scores)
+            mlvl_score_factor.append(score_factor)
+        return self._bbox_post_process(mlvl_scores, mlvl_bboxes,
```

Comment:
```
det_bboxes, det_labels = self._bbox_post_process(mlvl_scores, mlvl_bboxes,
                                       img_meta['scale_factor'], cfg, rescale,
                                       with_nms, mlvl_score_factor, **kwargs)
return det_bboxes, det_labels
```
may be clearer.

Reply:
we need consider
```sh
            if mlvl_score_factor is not None:
                return mlvl_bboxes, mlvl_scores, mlvl_score_factor
            else:
                return mlvl_bboxes, mlvl_scores
```

Reply:
You are right, but haven't found a better way for the time being

Reply:
I would refactor this in `Post-Process Piplines`